airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/default_file_based_stream.py

```diff
@@ -3,37 +3,63 @@
 #
 
 import asyncio
+import itertools
 import logging
-from datetime import datetime
+from datetime import timedelta
 from functools import cache
-from typing import Any, Iterable, List, Mapping, Optional, Union
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional, Union
 
-from airbyte_cdk.models import SyncMode
-from airbyte_cdk.sources.declarative.types import StreamSlice, StreamState
+from airbyte_cdk.sources.declarative.types import StreamSlice
 from airbyte_cdk.sources.file_based.exceptions import MissingSchemaError, RecordParseError, SchemaInferenceError
 from airbyte_cdk.sources.file_based.remote_file import RemoteFile
-from airbyte_cdk.sources.file_based.schema_helpers import merge_schemas, type_mapping_to_jsonschema
+from airbyte_cdk.sources.file_based.schema_helpers import merge_schemas
 from airbyte_cdk.sources.file_based.schema_validation_policies import record_passes_validation_policy
 from airbyte_cdk.sources.file_based.stream import AbstractFileBasedStream
+from airbyte_cdk.sources.file_based.stream.file_based_cursor import FileBasedCursor
+from airbyte_cdk.sources.streams import IncrementalMixin
 from airbyte_cdk.sources.utils.record_helper import stream_data_to_airbyte_message
 
 
-class DefaultFileBasedStream(AbstractFileBasedStream):
+class DefaultFileBasedStream(AbstractFileBasedStream, IncrementalMixin):
+
     """
     The default file-based stream.
     """
 
+    ab_last_mod_col = "_ab_source_file_last_modified"
+    ab_file_name_col = "_ab_source_file_url"
+    airbyte_columns = [ab_last_mod_col, ab_file_name_col]
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self._cursor = FileBasedCursor(
+            self.config.max_history_size or 10_000, timedelta(days=(self.config.days_to_sync_if_history_is_full or 3)), self.logger
+        )
+
+    @property
+    def state(self) -> MutableMapping[str, Any]:
+        return self._cursor.to_dict()
+
+    @state.setter
+    def state(self, value: MutableMapping[str, Any]):
+        """State setter, accept state serialized by state getter."""
+        self._cursor.set_initial_state(value)
+
     @property
     def primary_key(self) -> Optional[Union[str, List[str], List[List[str]]]]:
         return self.config.primary_key
 
-    def read_records(
-        self,
-        sync_mode: SyncMode,
-        cursor_field: List[str] = None,
-        stream_slice: Optional[StreamSlice] = None,
-        stream_state: Optional[StreamState] = None,
-    ) -> Iterable[Mapping[str, Any]]:
+    def compute_slices(self) -> Iterable[Optional[Mapping[str, Any]]]:
+        # Group all files by timestamps and return them as slices
+        # TODO: Partition the files by glob patterns to enable better checkpointing
```

Comment:
How would partitioning by glob pattern improve checkpointing?

Reply:
It would increase the checkpoint granularity.

Assuming we have two globs 
1. One glob with a single file for datetime = 2021-01-01T00:00:00 "*_a.csv" -> ["my_file_a.csv"]
2. One glob with many files for datetime = 2021-01-01T00:00:00 "*_b.csv" -> ["my_file_b.csv", "another_file_b.csv"]

We can only checkpoint the instant if all globs are synced. Checkpointing by glob would allow us to commit my_file_a.csv even we the _b files are large or if they fail.

The benefits are debatable since I wouldn't expect that many files having the exact same timestamp

Reply:
Thanks, I see what you mean. I suspect that the gains from partitioning by glob will be relatively low - just scrolling through [metabase](https://airbyte.metabaseapp.com/question#eyJkYXRhc2V0X3F1ZXJ5Ijp7InR5cGUiOiJuYXRpdmUiLCJuYXRpdmUiOnsicXVlcnkiOiJzZWxlY3RcbiAgICBDb3VudCgqKSxcbiAgICBjb25maWd1cmF0aW9uIC0-ICdwYXRoX3BhdHRlcm4nXG5mcm9tIGFjdG9yIHdoZXJlIGFjdG9yX2RlZmluaXRpb25faWQgPSAnNjk1ODk3ODEtNzgyOC00M2M1LTlmNjMtODkyNWIxYzFjY2MyJyBHUk9VUCBCWSBjb25maWd1cmF0aW9uIC0-ICdwYXRoX3BhdHRlcm4nOyIsInRlbXBsYXRlLXRhZ3MiOnt9fSwiZGF0YWJhc2UiOjcyfSwiZGlzcGxheSI6InRhYmxlIiwicGFyYW1ldGVycyI6W10sInZpc3VhbGl6YXRpb25fc2V0dGluZ3MiOnt9fQ==), most connections just have a single pattern (fyi globs will be `"|"`-separated when you look through the results).

Reply:
fair enough. doesn't seem too useful then